[Android][libs] Introduce GetLocalUtcOffset temporary fast result

mdh1418 opened the pull request:
Fixes https://github.com/dotnet/runtime/issues/71004

On Android, time zone data is loaded through AndroidTZData, and the time to load significantly impacts startup performance should one call an API reliant on `AndroidTzDataInstance` in the startup sequence.
e.g. Displaying `DateTimeOffset.Now` -> `TimeZoneInfo.GetLocalUtcOffset` -> `cachedData.Local`  -> `cachedData.CreateLocal` -> `GetLocalTimeZone` -> `GetLocalTimeZoneCore` -> `GetTimeZone` -> `AndroidTzDataInstance.GetTimeZoneData`

We can leverage Java API to obtain an offset that is good enough for `GetLocalUtcOffset` and pass that during the mono runtime startup via `monovm_initialize` or `monovm_initialize_preparsed` as a "fast path". In the background, we load AndroidTZData, as it is necessary for handling any time zone changes during the app's duration. 

------------------------------
# Testing through Android Sample App
```
Console.WriteLine($"The current time is `{DateTimeOffset.Now}`");
Thread.Sleep(1000);
var temp = DateTimeOffset.Now;
Console.WriteLine($"Now the current time is `{temp}`");
```

```
The current time is `8/23/2022 4:26:22 PM -04:00`
Now the current time is `8/23/2022 4:26:24 PM -04:00`
```

speedscope: https://gist.github.com/mdh1418/473d097cbd1d7b1187c9d83ddb8da11b

### First call to DateTimeOffset.Now (Fast Path) 145.66ms
![Screen Shot 2022-08-23 at 5 18 57 PM](https://user-images.githubusercontent.com/16830051/186267632-b0b2ecb6-37e0-47cc-bef1-d5d1d371eba4.png)

### Second call to DateTimeOffset.Now (After AndroidTZData loaded) 2.67ms
![Screen Shot 2022-08-23 at 5 19 44 PM](https://user-images.githubusercontent.com/16830051/186267918-9584b964-7b83-4269-a09b-739e8a488593.png)

------------------------------

# This PR does the following:
- Splits DateTimeOffset.Now into an Android and non Android implementation
- Modifies DateTimeOffset.Now on Android (1st call only) to start a background thread to load AndroidTZData and return a fast result passed in from Java side
- Introduces `System.TimeZoneInfo.LocalDateTimeOffset` property to be passed from Java either as an AppContext property or as a part of `MonoCoreRuntimeProperties`
- Avoid possible ICU loading akin to https://github.com/xamarin/java.interop/pull/1001/files

mdh1418 committed:
Avoid ICU loading via StringComparison.Ordinal

mdh1418 committed:
Pass OffsetDateTime and monotonic clock through monovm

mdh1418 committed:
WIP Leverage monovm_initialize_preparsed

mdh1418 committed:
Initial perf improvement implementation

mdh1418 committed:
Cleanup GetLocalUtcOffset

mdh1418 committed:
Remove Kernel Monotonic Time app context key

mdh1418 committed:
Cleanup GetLocalUtcOffset

mdh1418 commented:
@grendello @lambdageek Would it be preferred to pass the Offset obtained through Java API as an AppContext key/value property or as part of `MonoCoreRuntimeProperties`? If the latter, how should that be grabbed in the library side?

jonathanpeppers commented on src/libraries/System.Private.CoreLib/src/System/TimeZoneInfo.Unix.Android.cs:
```diff
@@ -142,6 +142,56 @@ private static TimeZoneInfo GetLocalTimeZoneCore()
+
+                    if (_loadAndroidTZData == null)
+                    {
+                        _loadAndroidTZData = new Thread(() => {
+                            _localUtcOffset = GetCacheLocalUtcOffset(dateTime, flags);
+                            Thread.Sleep(1000);
+                        });
+                        _loadAndroidTZData.IsBackground = true;
+                        _loadAndroidTZData.Start();
```
@grendello what are your thoughts here on starting a `Thread`?

grendello replied:
The thread idea was my suggestion and I think it's going to work quite OK, provided it's a bare thread as above and not part of a threadpool or TPL in general. It shouldn't hurt startup performance too much, although what I'd do would be to let it sleep **first** for 1 second and then start loading the TZ. The idea is that it should start after the main startup sequence of the app, and 1s should be enough time.

eerhardt replied:
If 1s isn't enough time, is this going to make launching the app even slower?

grendello commented:
> @grendello @lambdageek Would it be preferred to pass the Offset obtained through Java API as an AppContext key/value property or as part of `MonoCoreRuntimeProperties`? If the latter, how should that be grabbed in the library side?

I'd prefer it to be part of `MonoCoreRuntimeProperties`, it'd be faster to set and we can obtain the offset at startup from our Java code that invokes our netive runtime.

lambdageek commented:
> > @grendello @lambdageek Would it be preferred to pass the Offset obtained through Java API as an AppContext key/value property or as part of `MonoCoreRuntimeProperties`? If the latter, how should that be grabbed in the library side?
> 
> I'd prefer it to be part of `MonoCoreRuntimeProperties`, it'd be faster to set and we can obtain the offset at startup from our Java code that invokes our netive runtime.

I think it shouldn't be in `MonoCoreRuntimeProperties` 😁 

it's faster to set in XA, but in mono we will still have to convert it to a string property in order to populate the managed AppContext. So we're just pushing the cost down one layer

jonathanpeppers commented on src/libraries/System.Private.CoreLib/src/System/TimeZoneInfo.Unix.Android.cs:
```diff
@@ -142,6 +142,56 @@ private static TimeZoneInfo GetLocalTimeZoneCore()
+                    if (_loadAndroidTZData == null)
+                    {
+                        _loadAndroidTZData = new Thread(() => {
+                            _localUtcOffset = GetCacheLocalUtcOffset(dateTime, flags);
+                            Thread.Sleep(1000);
```
I think @grendello was suggesting to move the sleep up one line here.

mdh1418 replied:
I have the commit in a pending commit. I'll push it up, I was trying to bundle both this change and anything to do with parsed properties, but ill just separate.

grendello commented:
> > > @grendello @lambdageek Would it be preferred to pass the Offset obtained through Java API as an AppContext key/value property or as part of `MonoCoreRuntimeProperties`? If the latter, how should that be grabbed in the library side?
> > 
> > 
> > I'd prefer it to be part of `MonoCoreRuntimeProperties`, it'd be faster to set and we can obtain the offset at startup from our Java code that invokes our netive runtime.
> 
> I think it shouldn't be in `MonoCoreRuntimeProperties` grin
> 
> it's faster to set in XA, but in mono we will still have to convert it to a string property in order to populate the managed AppContext. So we're just pushing the cost down one layer

That's the whole point, though :) To make the startup sequence of XA faster - what happens after the app is running can be slower without hurting the perceived app performance.

lambdageek commented:
> That's the whole point, though :) To make the startup sequence of XA faster - what happens after the app is running can be slower without hurting the perceived app performance.

This won't make the startup faster.  We need to populate the AppContext during startup.

mdh1418 committed:
Sleep background thread first, then start loading AndroidTZData

lambdageek commented on src/native/public/mono/jit/details/mono-private-unstable-types.h:
```diff
@@ -45,10 +45,15 @@ typedef struct {
 	char **dirs;
 } MonoCoreLookupPaths;
 
+typedef struct {
+	uint64_t current_local_time;
+} MonoCoreLocalTime;
```
if we decide to use the preparsed structure,

1. it's not the _time_, it's the _UTC offset_, right?
2. we should add some comments to the header.

mdh1418 replied:
Ah right, the naming was prior to when I realized we needed to grab the offset itself. I'll fix it if we do end up leveraging the preparsed structure.

mdh1418 replied:
Removed the preparsed structure

lambdageek commented on src/libraries/System.Private.CoreLib/src/System/TimeZoneInfo.Unix.Android.cs:
```diff
@@ -142,6 +142,56 @@ private static TimeZoneInfo GetLocalTimeZoneCore()
+                }
+            }
+
+            object? localDateTimeOffset = AppContext.GetData("LOCAL_DATE_TIME_OFFSET");
```
Is there some prefix that we need to use? 
From the docs it looks like normally the names are namespaced like classes https://docs.microsoft.com/en-us/dotnet/core/runtime-config/

So maybe `System.TimeZoneInfo.LocalDataTimeOffset` ?

mdh1418 replied:
Thats a good point, as in the App Context Property's Key should follow that format for usage in the managed side? I think you're right, I was following convention that the other properties use, but then again, it doesn't seem like any of the other properties are directly used in the managed side. It will stand out here https://github.com/xamarin/xamarin-android/blob/a689fb9b0e1459c95115b4c9949cb7ec4068f2be/src/Xamarin.Android.Build.Tasks/Microsoft.Android.Sdk/targets/Microsoft.Android.Sdk.RuntimeConfig.targets#L23-L30, but the new naming will make it more clear where the property is being used right?

lambdageek replied:
@vitek-karas Do we keep track of important AppContext keys somewhere? Is there a convention for naming them?  Some way to avoid clashing with third-party names?

vitek-karas replied:
There's no convention which we all agreed on. For feature switches we use the "looks like managed namespace qualified name" - and if there's a managed counterpart (property) we typically use the same name. https://github.com/dotnet/runtime/blob/main/docs/workflow/trimming/feature-switches.md

Other knobs also tend to follow the "looks like managed" convention, but not all.

I think we should use that convention everywhere we can - just like namespaces it leads to pretty much guaranteed uniqueness. But that's just my opinion.

mdh1418 replied:
Named it to `System.TimeZoneInfo.LocalDateTimeOffset`

mdh1418 committed:
Conform Runtime Config Property name

mdh1418 commented:
/azp run runtime-extra-platforms

lambdageek reviewed (approved):
Let's not add the offset to the preinitialized properties right now.  Once that new field and typedef are removed, LGTM.  Let's try to backport this to rc2

eerhardt commented on src/libraries/System.Private.CoreLib/src/System/TimeZoneInfo.Unix.Android.cs:
```diff
@@ -107,7 +107,7 @@ private static int ParseGMTNumericZone(string name)
                 return new TimeZoneInfo(id, TimeSpan.FromSeconds(0), id, name, name, null, disableDaylightSavingTime:true);
             }
-            if (name.StartsWith("GMT", StringComparison.Ordinal))
+            if (name[0] == 'G' && name[1] == 'M' && name[2] == 'T')
```
Is there a guarantee that `name` is at least 3 characters long?

eerhardt commented on src/libraries/System.Private.CoreLib/src/System/TimeZoneInfo.Unix.Android.cs:
```diff
@@ -142,6 +142,56 @@ private static TimeZoneInfo GetLocalTimeZoneCore()
 
+        private static TimeSpan GetCacheLocalUtcOffset(DateTime dateTime, TimeZoneInfoOptions flags)
+        {
+                CachedData cachedData = s_cachedData;
+                return cachedData.Local.GetUtcOffset(dateTime, flags, cachedData);
```
```suggestion
            CachedData cachedData = s_cachedData;
            return cachedData.Local.GetUtcOffset(dateTime, flags, cachedData);
```

mdh1418 replied:
Thanks, forgot to unindent the paste!

eerhardt commented on src/libraries/System.Private.CoreLib/src/System/TimeZoneInfo.Unix.Android.cs:
```diff
@@ -142,6 +142,56 @@ private static TimeZoneInfo GetLocalTimeZoneCore()
+                return cachedData.Local.GetUtcOffset(dateTime, flags, cachedData);
+        }
+
+        private static TimeSpan? _localUtcOffset;
+        private static object _localUtcOffsetLock = new();
+        private static Thread? _loadAndroidTZData;
```
```suggestion
        private static TimeSpan? s_localUtcOffset;
        private static object s_localUtcOffsetLock = new();
        private static Thread? s_loadAndroidTZData;
```

marek-safar commented on src/libraries/System.Private.CoreLib/src/System/TimeZoneInfo.Unix.Android.cs:
```diff
@@ -142,6 +142,56 @@ private static TimeZoneInfo GetLocalTimeZoneCore()
+                            _localUtcOffset = GetCacheLocalUtcOffset(dateTime, flags);
+                        });
+                        _loadAndroidTZData.IsBackground = true;
+                        _loadAndroidTZData.Start();
```
Who is going to remove/clear the thread?

mdh1418 replied:
Whats the best way to do that? Is the way to remove/clear Threads to set `_loadAndroidTZData` to `null` once we know the thread is finished? (maybe in the first `if (_localUtcOffset != null) // The background thread finished, the cache is loaded.`)

mdh1418 replied:
Setting `s_loadAndroidTZData` to null when we know AndroidTZData is loaded (`s_androidTZDataLoaded == true`)

eerhardt commented on src/libraries/System.Private.CoreLib/src/System/TimeZoneInfo.Unix.Android.cs:
```diff
@@ -142,6 +142,56 @@ private static TimeZoneInfo GetLocalTimeZoneCore()
+                return cachedData.Local.GetUtcOffset(dateTime, flags, cachedData);
+        }
+
+        private static TimeSpan? _localUtcOffset;
```
Does this actually need to be a `TimeSpan`? It doesn't seem like we are using this for anything other than setting the "we've loaded the TZ Data". Can it just be a boolean?

mdh1418 replied:
Right, we don't leverage the value besides checking that it is set. I can modify `s_localUtcOffset` to be a boolean, and modify the thread body to be 
```
Thread.Sleep(1000);
_ = GetCacheLocalUtcOffset(dateTime, flags);
s_localUtcOffset = true;
```
Or perhaps instead of calling `GetCacheLocalUtcOffset`, the thread might only need to initialize AndroidTZData/invoke `cachedData.Local`?

mdh1418 replied:
Changed the variable to be a boolean that is set to true at the end of the background thread's body, and instead of calling `GetCacheLocalUtcOffset`, just invoked `cachedData.Local` to ensure the s_cachedData is populated, which should eventually load AndroidTZData.

eerhardt commented on src/libraries/System.Private.CoreLib/src/System/TimeZoneInfo.Unix.Android.cs:
```diff
@@ -142,6 +142,56 @@ private static TimeZoneInfo GetLocalTimeZoneCore()
+            if (localDateTimeOffset == null)
+                return GetCacheLocalUtcOffset(dateTime, flags); // If no offset property provided through monovm app context, default
+
+            long localDateTimeOffsetSeconds = Convert.ToInt32(localDateTimeOffset);
```
Why both `long` and `Int32` in this line? Shouldn't this either be:

```suggestion
            int localDateTimeOffsetSeconds = Convert.ToInt32(localDateTimeOffset);
```
or
```suggestion
            long localDateTimeOffsetSeconds = Convert.ToInt64(localDateTimeOffset);
```

mdh1418 replied:
Thanks for the catch, it slipped in when I was removing an in between variable. I think we can keep it as an `Int32` as the value passed from Java is an int and the largest an offset *should* be -14 hours from UTC (-50400)

jonathanpeppers commented on src/libraries/System.Private.CoreLib/src/System/TimeZoneInfo.Unix.Android.cs:
```diff
@@ -142,6 +142,56 @@ private static TimeZoneInfo GetLocalTimeZoneCore()
+                    if (_loadAndroidTZData == null)
+                    {
+                        _loadAndroidTZData = new Thread(() => {
+                            Thread.Sleep(1000);
```
Like @eerhardt mentioned the 1 second here, a slow app could be more than 1 second, is there a way this can use `ManualResetEvent`? (or whatever the "good" reset event is)

Then something in xamarin/xamarin-android, maybe `Android.App.Activity.OnCreate()` could call into the runtime. @grendello do you have any thoughts how we could signal to the runtime the app is done loading?

grendello replied:
If the app takes more than 1s to start, then the time to load TZ doesn't really matter. There's no reason to complicate the code with signalling, especially that Android is notoriously hard in this respect... The app is done loading when the `Displayed: app` line is output to logcat, but there's no way to catch that event in the app itself. 

1s is very generous, but we can increase it to 1.5s if we think this is better. However, anything above ~700ms is perceived as "slow" by the user, so I really wouldn't worry about this.

mdh1418 committed:
Revert "WIP Leverage monovm_initialize_preparsed"

This reverts commit 1edaeeee37066ad392ab19100248994ee53b0797.

mdh1418 committed:
Address feedback

mdh1418 commented:
I tested the fast path fallback, (i.e. making `localDateTimeOffset` null) and it behaves as expected.

The latest speedscopes I obtained locally https://gist.github.com/mdh1418/3dca442e80585b2da772399c1ff36fd1
Initial call to `DateTimeOffset.Now` 128.26ms
<img width="2493" alt="Screen Shot 2022-08-25 at 12 35 48" src="https://user-images.githubusercontent.com/16830051/186721392-74780b69-aabb-4e33-a245-520513561756.png">
Second call to `DateTimeOffset.Now` after background thread finished 26.07ms
<img width="2496" alt="Screen Shot 2022-08-25 at 12 36 03" src="https://user-images.githubusercontent.com/16830051/186721391-331ac580-25bb-4fd7-8760-50061e2a491c.png">

mdh1418 commented:
Running this sample 11 times in Release mode (all previous timings were in Debug)
```
Console.WriteLine($"The current time is `{DateTimeOffset.Now}`");
Thread.Sleep(1500);
var temptwo = DateTimeOffset.Now;
Console.WriteLine($"The current time is now `{temptwo}`");
```

The average timing of the initial call to DateTimeOffset.Now is **74.85ms**
The average timing of another call to DateTimeOffset.Now after the background thread is finished is **25.52ms**
<img width="1134" alt="Screen Shot 2022-08-25 at 15 14 28" src="https://user-images.githubusercontent.com/16830051/186749424-c18499a5-26e6-4d9b-9277-a0ba41876c31.png">

mdh1418 commented:
/azp run runtime-extra-platforms

mdh1418 committed:
Fix thread clearing

mdh1418 commented:
/azp run runtime-extra-platforms

eerhardt commented on src/tasks/AndroidAppBuilder/Templates/MonoRunner.java:
```diff
@@ -88,7 +90,8 @@ public static int initialize(String entryPointLibName, String[] args, Context co
 
         Log.i("DOTNET", "MonoRunner initialize,, entryPointLibName=" + entryPointLibName);
-        return initRuntime(filesDir, cacheDir, testResultsDir, entryPointLibName, args);
+        int localDateTimeOffset = OffsetDateTime.now().getOffset().getTotalSeconds();
```
Is there no way in Java to get the current TZ offset without calculating the current time?

eerhardt replied:
One issue is that all Android apps will be paying this cost of `OffsetDateTime.now()` here, even if the app doesn't use `DateTime.Now` in C# - especially in startup code.

Is there any "fast path" Java/Android API we could P/Invoke into in `GetLocalUtcOffset` instead?

eerhardt replied:
Actually, looking into this more, I'm not sure how this is correct to use `.now()` in the `TimeZoneInfo.GetLocalUtcOffset` method at all.

`TimeZoneInfo.GetLocalUtcOffset` takes the `DateTime` that we are interested in getting the offset for. Imagine if the first call passed in a `DateTime` from the winter and the app is currently running in the summer. That first call (for the first 1 seconds) is going to return the incorrect value. It is going to return an offset with daylight savings time, but the `DateTime` being requested isn't under daylight savings time.

mdh1418 replied:
I think you're right. Since this performance issue was a problem with AndroidTZData taking a while to load, I tried to bump the implementation difference further down trying to account for all APIs that invoke AndroidTZData. However, it was decided offline that we can focus on `DateTimeOffset.Now` as it is more likely to be called at startup (compared to others), and other APIs would require a different information passed from Java to the mono runtime. Along the way I didn't pay enough attention to that `DateTimeOffset.Now` only calls one specific case of `GetLocalUtcOffset` when trying to get the background thread working. 

Since this PR is aimed at resolving the performance for `DateTimeOffset.Now`, would it be better to migrate `Now` into new `DateTimeOffset.cs` `DateTimeOffset.NonAndroid.cs` files and introduce the logic of fast path + background thread for the offset in there?

mdh1418 replied:
Yeah, I'm realizing now that I conflated two different ideas from the original issue. I think @grendello mentioned that a P/Invoke + JNI would work, but might be slightly slower than the original ideas in https://github.com/dotnet/runtime/issues/71004#issuecomment-1172131137

grendello replied:
It would be enough if dotnet runtime handled the timezone change broadcast - or used one that [Xamarin.Android installs](https://github.com/xamarin/xamarin-android/blob/91f330f4a57af90c5f7bde223cf204b6afb5324f/src/java-runtime/java/mono/android/MonoPackageManager.java#L33-L36). My original suggestion was that Xamarin.Android would get the local offset somewhere [around here](https://github.com/xamarin/xamarin-android/blob/91f330f4a57af90c5f7bde223cf204b6afb5324f/src/java-runtime/java/mono/android/MonoPackageManager.java#L38-L44) and pass it to the Mono runtime when we call `monovm_inititialize_preparsed`. **If** within that time the tz would have changed, we would be able to capture that and pass the changed value instead. If the tz change happens after we call `monovm_initialize_preparsed`, then the notification broadcast handler would catch it and advise MonoVM about that (e.g. by calling a hypothetical `mono_update_tz_offset` function). But since that idea wasn't implemented, the next best thing is for MonoVM add its own tz change notification handler a'la the one in Xamarin.Android. It would be a bit wasteful to have two of them, but it would seem that currently it would be the only solution.

eerhardt replied:
My comment isn't about the time zone changing. It is referring to the `DateTime` that is being passed into `GetLocalUtcOffset`. The method being changed here is called from many places, not just `DateTimeOffset.Now`.

One example is `DateTimeOffset.ToLocalTime()`. If I was to run the following code:

```C#
var localTime = new DateTimeOffset(
    new DateTime(2022, 12, 29, 12, 0, 0),
    TimeSpan.Zero)
    .ToLocalTime();

Console.WriteLine(localTime.LocalDateTime);
```

That would produce different results based on _when_ the app was executed. If I ran the app in the summer, it would produce (for me in US Central):

```
12/29/2022 6:00:00 AM
```

If I ran it in the winter, it would produce:

```
12/29/2022 7:00:00 AM
```

Because we would be using the offset for `DateTime.Now` in that first second, which is not correct.

mdh1418 replied:
> Is there no way in Java to get the current TZ offset without calculating the current time?

When I looked through the docs, it didn't seem like there was a direct "Current System Time Offset To UTC" API to leverage. There were multiple candidates for obtaining the value, such as `OffsetDateTime.now().getOffset().getTotalSeconds();` and `ZonedDateTime.now().getOffset().getTotalSeconds();`, but from timing via `Instant.now()` and `System.currentTimeMillis()` they seem to be similar (measuring 0 time difference). 


> One issue is that all Android apps will be paying this cost of OffsetDateTime.now() here, even if the app doesn't use DateTime.Now in C# - especially in startup code.

I think alignment with what Xamarin-Android does (which will be paying the cost of grabbing the local UTC offset via Java APIs) is preferred over the cost, plus the cost seems small.

eerhardt commented on src/libraries/System.Private.CoreLib/src/System/TimeZoneInfo.Unix.Android.cs:
```diff
@@ -142,6 +142,62 @@ private static TimeZoneInfo GetLocalTimeZoneCore()
+        }
+
+        private static bool s_androidTZDataLoaded;
+        private static object s_localUtcOffsetLock = new();
```
Add `readonly`

mdh1418 committed:
Move implementation to DateTimeOffset.Now

eerhardt commented on src/libraries/System.Private.CoreLib/src/System/DateTimeOffset.NonAndroid.cs:
```diff
@@ -0,0 +1,12 @@
+
+namespace System
+{
+        public readonly partial struct DateTimeOffset
+        {
+                // Returns a DateTimeOffset representing the current date and time. The
+                // resolution of the returned value depends on the system timer.
+                public static DateTimeOffset Now => ToLocalTime(DateTime.UtcNow, true);
+        }
```
```suggestion
    public readonly partial struct DateTimeOffset
    {
        // Returns a DateTimeOffset representing the current date and time. The
        // resolution of the returned value depends on the system timer.
        public static DateTimeOffset Now => ToLocalTime(DateTime.UtcNow, true);
    }
```

too much indention

eerhardt commented on src/libraries/System.Private.CoreLib/src/System/DateTimeOffset.Android.cs:
```diff
@@ -0,0 +1,80 @@
+
+        private static TimeSpan TryGetNow()
+        {
+            if (s_androidTZDataLoaded) // The background thread finished, the cache is loaded.
+                return TimeZoneInfo.GetLocalUtcOffset(DateTime.UtcNow, TimeZoneInfoOptions.NoThrowOnInvalidTime);
```
Why aren't we using the same code as the "NonAndroid" path when `s_androidTZDataLoaded == true`?

Basically, in `DateTimeOffset.Now`:

```C#
            if (s_androidTZDataLoaded) // The background thread finished, the cache is loaded.
                return ToLocalTime(DateTime.UtcNow, true);

            // Kick off the background thread, if necessary and return the fast path
````

eerhardt replied:
The issue is we are calling `DateTime.UtcNow` multiple times in the "happy" path - once in the `Now` property and once here.

mdh1418 replied:
I wanted this helper function to just provide a fast path for the offset, essentially just branching out https://github.com/dotnet/runtime/blob/f825814b3561c5c7aa8d66773b338944c2c72a3c/src/libraries/System.Private.CoreLib/src/System/DateTimeOffset.cs#L823.

The `DateTimeOffset.Now` below is essentially `ToLocalTime(DateTime.UtcNow, true)`.

I thought it was more direct to specify this fast path behavior for `DateTimeOffset.Now`. Otherwise, if we had to go one more level and specify it for `ToLocalTime(DateTime, bool)` then it would be like the situation we had for `TimeZoneOffset.GetLocalUtcOffset()` where it special behavior for a specific set of parameters.

mdh1418 replied:
Would this be preferred
```
        public static DateTimeOffset Now
        {
            get
            {
                if (s_androidTZDataLoaded) // The background thread finished, the cache is loaded.
                    return ToLocalTime(DateTime.UtcNow, true);

                if (!s_androidTZDataLoaded && s_loadAndroidTZData == null) // The cache isn't loaded and no background thread has been created
                {
                    lock (s_localUtcOffsetLock)
                    {
                        // Now may be called multiple times before a cache is loaded and a background thread is running,
                        // once the lock is available, check for a cache and background thread.
                        if (s_androidTZDataLoaded)
                            return ToLocalTime(DateTime.UtcNow, true);

                        if (s_loadAndroidTZData == null)
                        {
                            s_loadAndroidTZData = new Thread(() => {
                                // Delay the background thread to avoid impacting startup, if it still coincides, startup is already perceived as slow
                                Thread.Sleep(1000);
                                _ = TimeZoneInfo.Local; // Load AndroidTZData
                                lock (s_localUtcOffsetLock)
                                {
                                    s_androidTZDataLoaded = true;
                                    s_loadAndroidTZData = null; // Ensure thread is cleared when cache is loaded
                                }
                            });
                            s_loadAndroidTZData.IsBackground = true;
                            s_loadAndroidTZData.Start();
                        }
                    }
                }

                object? localDateTimeOffset = AppContext.GetData("System.TimeZoneInfo.LocalDateTimeOffset");
                if (localDateTimeOffset == null) // If no offset property provided through monovm app context, default
                    return ToLocalTime(DateTime.UtcNow, true);

                // Fast path obtained offset into ToLocalTime(DateTime.UtcNow, true) logic
                DateTime utcDateTime = DateTime.UtcNow;
                int localDateTimeOffsetSeconds = Convert.ToInt32(localDateTimeOffset);
                TimeSpan offset = TimeSpan.FromSeconds(localDateTimeOffsetSeconds);
                long localTicks = utcDateTime.Ticks + offset.Ticks;
                if (localTicks < DateTime.MinTicks || localTicks > DateTime.MaxTicks)
                    throw new ArgumentException(SR.Arg_ArgumentOutOfRangeException);

                return new DateTimeOffset(localTicks, offset);
            }
        }
```

mdh1418 replied:
I consolidated the logic into just `DateTimeOffset.Now`, and passed around a var to avoid calling `DateTime.UtcNow` twice.

eerhardt commented on src/libraries/System.Private.CoreLib/src/System/DateTimeOffset.Android.cs:
```diff
@@ -0,0 +1,80 @@
+            get
+            {
+                DateTime utcDateTime = DateTime.UtcNow;
+                TimeSpan offset = TryGetNow();
```
Is there a better name than `TryGetNow`? Maybe `TryGetNowOffset()`?

mdh1418 replied:
The name is flexible, I vaguely recall seeing somewhere else for a method helper having the `Try` preceding the name so I just started off with that, but that might have had a `try/catch` involved. In any case its just a helper to try to get the fast path offset.

mdh1418 replied:
Consolidated all the logic (fast path + threading) into `DateTimeOffset.Now` rather than in just trying to get its offset, so theres no helper function now.

eerhardt commented on src/tasks/AndroidAppBuilder/Templates/monodroid.c:
```diff
@@ -225,13 +225,17 @@ mono_droid_runtime_init (const char* executable, int managed_argc, char* managed
+    const char* appctx_keys[3];
     appctx_keys[0] = "RUNTIME_IDENTIFIER";
     appctx_keys[1] = "APP_CONTEXT_BASE_DIRECTORY";
+    appctx_keys[2] = "System.TimeZoneInfo.LocalDateTimeOffset";
```
Are we considering this as "internal-use only"? Or would we expect developers to get this AppContext setting?

With its current name, it kind of seems like an "official" setting. I wonder if we should rename it to better indicate that we don't expect others to use it. (and quite possibly we could remove it in the future if we refactor this code.)

mdh1418 replied:
I believe we want to set the MonoVMProperty on Xamarin-Android at all times. And to address https://github.com/dotnet/runtime/pull/74459#discussion_r956348984, I believe we would want to maintain alignment of AndroidAppBuilder in `dotnet/runtime` with what is done in Xamarin-Android over any blanket cost that might be added to all apps. However, I did some simple timing using `System.currentTimeMillis()` and `Instant.now()` and it seems instantaneous.

I think the only purpose for this property is to get the fast path local date time offset, but there was a concern whether we needed this to follow runtime config naming https://github.com/dotnet/runtime/pull/74459#discussion_r953947814.

mdh1418 committed:
Address feedback

mdh1418 commented:
/azp run runtime-extra-platforms

mdh1418 committed:
Consolidate into DateTimeOffset.Now and avoid calling DateTime.UtcNow multiple times

mdh1418 commented:
/azp run runtime-extra-platforms

eerhardt commented on src/libraries/System.Private.CoreLib/src/System/DateTimeOffset.Android.cs:
```diff
@@ -0,0 +1,76 @@
+
+                if (!s_androidTZDataLoaded && s_loadAndroidTZData == null) // The cache isn't loaded and no background thread has been created
+                {
+                    lock (s_localUtcOffsetLock)
+                    {
+                        // Now may be called multiple times before a cache is loaded and a background thread is running,
+                        // once the lock is available, check for a cache and background thread.
+                        if (s_androidTZDataLoaded)
+                            return ToLocalTime(utcDateTime, true);
+
+                        if (s_loadAndroidTZData == null)
+                        {
+                            s_loadAndroidTZData = new Thread(() => {
+                                // Delay the background thread to avoid impacting startup, if it still coincides after 1s, startup is already perceived as slow
+                                Thread.Sleep(1000);
+                                _ = TimeZoneInfo.Local; // Load AndroidTZData
+                                lock (s_localUtcOffsetLock)
+                                {
+                                    s_androidTZDataLoaded = true;
+                                    s_loadAndroidTZData = null; // Ensure thread is cleared when cache is loaded
+                                }
+                            });
+                            s_loadAndroidTZData.IsBackground = true;
+                            s_loadAndroidTZData.Start();
+                        }
+                    }
+                }
+
+                object? localDateTimeOffset = AppContext.GetData("System.TimeZoneInfo.LocalDateTimeOffset");
+                if (localDateTimeOffset == null) // If no offset property provided through monovm app context, default
+                    return ToLocalTime(utcDateTime, true);
+
+                // Fast path obtained offset incorporated into ToLocalTime(DateTime.UtcNow, true) logic
+                int localDateTimeOffsetSeconds = Convert.ToInt32(localDateTimeOffset);
+                TimeSpan offset = TimeSpan.FromSeconds(localDateTimeOffsetSeconds);
+                long localTicks = utcDateTime.Ticks + offset.Ticks;
+                if (localTicks < DateTime.MinTicks || localTicks > DateTime.MaxTicks)
+                    throw new ArgumentException(SR.Arg_ArgumentOutOfRangeException);
+
+                return new DateTimeOffset(localTicks, offset);
```
(question) Is there any performance benefit into making this a separate method since after the first second it will no longer be hit?

mdh1418 replied:
I'm not sure if I'm understanding your question correctly, but is the question whether or not something like this
```
        private static DateTimeOffset FastPathHelper(DateTime utcDateTime)
        {
            if (!s_androidTZDataLoaded && s_loadAndroidTZData == null) // The cache isn't loaded and no background thread has been created
            {
                lock (s_localUtcOffsetLock)
                {
                    // Now may be called multiple times before a cache is loaded and a background thread is running,
                    // once the lock is available, check for a cache and background thread.
                    if (s_androidTZDataLoaded)
                        return ToLocalTime(utcDateTime, true);

                    if (s_loadAndroidTZData == null)
                    {
                        s_loadAndroidTZData = new Thread(() => {
                            // Delay the background thread to avoid impacting startup, if it still coincides after 1s, startup is already perceived as slow
                            Thread.Sleep(1000);
                            _ = TimeZoneInfo.Local; // Load AndroidTZData
                            lock (s_localUtcOffsetLock)
                            {
                                s_androidTZDataLoaded = true;
                                s_loadAndroidTZData = null; // Ensure thread is cleared when cache is loaded
                            }
                        });
                        s_loadAndroidTZData.IsBackground = true;
                        s_loadAndroidTZData.Start();
                    }
                }
            }

            object? localDateTimeOffset = AppContext.GetData("System.TimeZoneInfo.LocalDateTimeOffset");
            if (localDateTimeOffset == null) // If no offset property provided through monovm app context, default
                return ToLocalTime(utcDateTime, true);

            // Fast path obtained offset incorporated into ToLocalTime(DateTime.UtcNow, true) logic
            int localDateTimeOffsetSeconds = Convert.ToInt32(localDateTimeOffset);
            TimeSpan offset = TimeSpan.FromSeconds(localDateTimeOffsetSeconds);
            long localTicks = utcDateTime.Ticks + offset.Ticks;
            if (localTicks < DateTime.MinTicks || localTicks > DateTime.MaxTicks)
                throw new ArgumentException(SR.Arg_ArgumentOutOfRangeException);

            return new DateTimeOffset(localTicks, offset);
        }
        
        public static DateTimeOffset Now
        {
            get
            {
                DateTime utcDateTime = DateTime.UtcNow;

                if (s_androidTZDataLoaded) // The background thread finished, the cache is loaded.
                    return ToLocalTime(utcDateTime, true);

                return FastPathHelper(utcDateTime);
            }
        }
        ```
Is faster?

eerhardt replied:
Correct - that is the question. Are there any perf benefits to splitting the property up that way?

mdh1418 replied:
I dont see any obvious performance gains, and I spoke with @grendello offline. From what I understand there would be more JIT time if it was split (JIT time for the new method and a call to it) than the current JIT time for just the getter and no call to a new method.

eerhardt replied:
Thanks for the confirmation!

eerhardt reviewed (approved):
This looks good to me. I just have a minor question about splitting the property getter.

mdh1418 committed:
Introduce small performance gains

Leverage a boolean flag instead of &&
Start the background thread to avoid deadlocking

mdh1418 commented:
@grendello, I tweaked a little bit to get the thread starting after the lock to work. Without a condition, it could crash with some test like
```
Console.WriteLine($"The current time is `{DateTimeOffset.Now}`");
var test = DateTimeOffset.Now;
Console.WriteLine($"Now the current time is `{test}`");
Thread.Sleep(1000);
var hwang = DateTimeOffset.Now;
Console.WriteLine($"Now the current time is `{hwang}`");
Thread.Sleep(500);
var temp = DateTimeOffset.Now;
Console.WriteLine($"Now the current time is `{temp}`");
```

mdh1418 commented:
/azp run runtime-extra-platforms

mdh1418 committed:
Merge remote-tracking branch 'upstream/main' into android_tzdata_perf_improvement

mdh1418 commented:
/azp run runtime-extra platforms

mdh1418 commented:
Checks didn't run because of the migration, should run this time after merging main

mdh1418 committed:
Bump boolean flag outside of background thread body

mdh1418 commented:
/azp run runtime-extra platforms

mdh1418 commented:
https://github.com/dotnet/runtime/issues/74952
https://github.com/dotnet/runtime/issues/71887
https://github.com/dotnet/runtime/issues/74896

The Android libraries tests didnt have any failures in a previous CI build, and the commits since then were only small performance gain modifications.

mdh1418 commented:
/backport to release/7.0

mdh1418 commented:
/backport to release/7.0
